igvc_gazebo/nodes/ground_truth/main.cpp

```diff
@@ -1,32 +1,38 @@
+#include <functional>
 #include <geometry_msgs/Vector3.h>
+#include <iterator>
+#include <mutex>
 #include <nav_msgs/Odometry.h>
 #include <parameter_assertions/assertions.h>
+#include <random>
 #include <robot_localization/navsat_conversions.h>
 #include <ros/ros.h>
 #include <tf/transform_broadcaster.h>
 #include <tf/transform_datatypes.h>
 #include <tf/transform_listener.h>
-#include <mutex>
 
 ros::Publisher g_ground_truth_pub;
 // TODO make this a minimal object
 nav_msgs::Odometry g_og_pose;
 ros::Time g_last_estimate;
+double noise;
 
-void odomCallback(const nav_msgs::Odometry::ConstPtr& msg)
+void odomCallback(const nav_msgs::Odometry::ConstPtr &msg)
 {
   g_last_estimate = msg->header.stamp;
 }
 
-void groundTruthCallback(const nav_msgs::Odometry::ConstPtr& msg)
+void groundTruthCallback(const nav_msgs::Odometry::ConstPtr &msg)
 {
+  // random noise generator
+  auto dist = std::bind(std::normal_distribution<double>{ 0.0, noise }, std::mt19937(std::random_device{}()));
   // get the starting location as the origin
   if (g_og_pose.header.stamp.toSec() == 0)
   {
     g_og_pose.pose = msg->pose;
```

Comment:
With the comment on having different noise variances for different fields, you can add noise to the heading by accessing the `orientation` member here. Remember to use the `tf::getYaw` and `tf::createQuaternionMsgFromYaw` helpers so that you can convert to / from quaternions.